pkg/compose/build_buildkit.go

```diff
@@ -71,3 +76,214 @@ func (s *composeService) doBuildBuildkit(ctx context.Context, project *types.Pro
+		},
+	}, nil
+}
+
+// driversForNodeGroup returns drivers for a nodegroup instance
+func driversForNodeGroup(ctx context.Context, dockerCli command.Cli, ng *store.NodeGroup, contextPathHash string) ([]build.DriverInfo, error) {
+	eg, _ := errgroup.WithContext(ctx)
+
+	dis := make([]build.DriverInfo, len(ng.Nodes))
+
+	var f driver.Factory
+	if ng.Driver != "" {
+		f = driver.GetFactory(ng.Driver, true)
+		if f == nil {
+			return nil, errors.Errorf("failed to find driver %q", f)
+		}
+	} else {
+		dockerapi, err := clientForEndpoint(dockerCli, ng.Nodes[0].Endpoint)
+		if err != nil {
+			return nil, err
+		}
+		f, err = driver.GetDefaultFactory(ctx, dockerapi, false)
+		if err != nil {
+			return nil, err
+		}
+		ng.Driver = f.Name()
+	}
+	imageopt, err := storeutil.GetImageConfig(dockerCli, ng)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, n := range ng.Nodes {
+		func(i int, n store.Node) {
+			eg.Go(func() error {
+				di := build.DriverInfo{
+					Name:        n.Name,
+					Platform:    n.Platforms,
+					ProxyConfig: storeutil.GetProxyConfig(dockerCli),
+				}
+				defer func() {
+					dis[i] = di
+				}()
+				dockerapi, err := clientForEndpoint(dockerCli, n.Endpoint)
+				if err != nil {
+					di.Err = err
+					return nil
+				}
+				// TODO: replace the following line with dockerclient.WithAPIVersionNegotiation option in clientForEndpoint
+				dockerapi.NegotiateAPIVersion(ctx)
+
+				contextStore := dockerCli.ContextStore()
+
+				var kcc driver.KubeClientConfig
+				kcc, err = configFromContext(n.Endpoint, contextStore)
+				if err != nil {
+					// err is returned if n.Endpoint is non-context name like "unix:///var/run/docker.sock".
+					// try again with name="default".
+					// FIXME: n should retain real context name.
+					kcc, err = configFromContext("default", contextStore)
+					if err != nil {
+						logrus.Error(err)
+					}
+				}
+
+				tryToUseKubeConfigInCluster := false
+				if kcc == nil {
+					tryToUseKubeConfigInCluster = true
+				} else {
+					if _, err := kcc.ClientConfig(); err != nil {
+						tryToUseKubeConfigInCluster = true
+					}
+				}
+				if tryToUseKubeConfigInCluster {
+					kccInCluster := driver.KubeClientConfigInCluster{}
+					if _, err := kccInCluster.ClientConfig(); err == nil {
+						logrus.Debug("using kube config in cluster")
+						kcc = kccInCluster
+					}
+				}
+
+				d, err := driver.GetDriver(ctx, "buildx_buildkit_"+n.Name, f, dockerapi, imageopt.Auth, kcc, n.Flags, n.Files, n.DriverOpts, n.Platforms, contextPathHash)
+				if err != nil {
+					di.Err = err
+					return nil
+				}
+				di.Driver = d
+				di.ImageOpt = imageopt
+				return nil
+			})
+		}(i, n)
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	return dis, nil
+}
+
+func configFromContext(endpointName string, s ctxstore.Reader) (clientcmd.ClientConfig, error) {
+	if strings.HasPrefix(endpointName, "kubernetes://") {
+		u, _ := url.Parse(endpointName)
```

Comment:
Wondering if we should have this in Compose, now that support for kubernetes endpoints having been removed from the docker CLI

(same for registering the kubernetes driver)

Reply:
The `kubernetes` endpoints were added to `docker context` in support of compose-on-kubernetes, which was deprecated and removed (together with support for creating contexts that use kubernetes endpoints), so I wonder if support for kubernetes should be added here (knowing that at least the contexts can no longer be created, but also that the rest of the CLI has no kubernetes support).

The scope of buildx is somewhat different, as it can also be used  without docker (which is mostly where these things come in)